### PR TITLE
Update 1.2.9 - Memory Leaks & Improvements

### PR DESCRIPTION
**Pull request by kavinprabhuc**
- Removed global view initialisation.
- Make use of onViewCreate lifecycle callback.

- Convert PaintFragment.java to Kotlin
- Introduce PaintViewModel.kt to fix leaks issue caused by BrushConfigDialog & EraserConfigDialog
- Removed kotlin std lib as Kotlin 1.4 by default has this included
- Refactor of Sticker and Paint fragments
 
- Updated dependencies and gradle
 
- Added a companion object fun in ImageEditorIntentBuilder.kt for simple intent creation with all features

### CODE REVIEW

File: ananas/src/main/java/iamutkarshtiwari/github/io/ananas/editimage/ImageEditorIntentBuilder.kt [hunk 1]
```diff
@@ -141,5 +141,37 @@ class ImageEditorIntentBuilder @JvmOverloads constructor(private val context: Co
...
+                    .withAddText() // Add the features you need
+                    .withPaintFeature()
+                    .withFilterFeature()
+                    .withRotateFeature()
+                    .withCropFeature()
+                    .withBrightnessFeature()
+                    .withSaturationFeature()
+                    .withBeautyFeature()
+                    .withStickerFeature()
+                    .forcePortrait(true) // Add this to force portrait mode (It's set to false by default)
```
Comment (pal-tgiannakopoulos) [inline ananas/src/main/java/iamutkarshtiwari/github/io/ananas/editimage/ImageEditorIntentBuilder.kt]: why do we need this?

File: ananas/src/main/java/iamutkarshtiwari/github/io/ananas/editimage/ImageEditorIntentBuilder.kt [hunk 2]
```diff
@@ -141,5 +141,37 @@ class ImageEditorIntentBuilder @JvmOverloads constructor(private val context: Co
...
```
Comment (pal-tgiannakopoulos) [inline ananas/src/main/java/iamutkarshtiwari/github/io/ananas/editimage/ImageEditorIntentBuilder.kt]: similar, why?

File: ananas/src/main/java/iamutkarshtiwari/github/io/ananas/editimage/ImageEditorIntentBuilder.kt [hunk 3]
```diff
@@ -141,5 +141,37 @@ class ImageEditorIntentBuilder @JvmOverloads constructor(private val context: Co
         const val FORCE_PORTRAIT = "force_portrait"
         const val EDITOR_TITLE = "editor_title"
         const val SUPPORT_ACTION_BAR_VISIBILITY = "support_action_bar_visibility"
+
+        fun getImageEditorIntent(context: Context, imageLocation: Uri, outputPath: String): Intent {
```
Comment (pal-tgiannakopoulos) [inline ananas/src/main/java/iamutkarshtiwari/github/io/ananas/editimage/ImageEditorIntentBuilder.kt]: better for these functions to be on our code to be more easily changeable

File: ananas/src/main/java/iamutkarshtiwari/github/io/ananas/editimage/adapter/StickerAdapter.java [hunk 4]
```diff
@@ -68,7 +68,12 @@ public void addStickerImages(String folderPath, int stickerCount) {
         @Override
         public void onClick(View v) {
             String data = (String) v.getTag();
-            stickerFragment.selectedStickerItem(data);
+            mCallback.onStickerSelected(data);
+//            stickerFragment.selectedStickerItem(data);
```
Comment (pal-tgiannakopoulos) [inline ananas/src/main/java/iamutkarshtiwari/github/io/ananas/editimage/adapter/StickerAdapter.java]: remove comment line pls

File: ananas/src/main/java/iamutkarshtiwari/github/io/ananas/editimage/fragment/BaseEditFragment.java [hunk 5]
```diff
@@ -1,23 +1,27 @@
...
+
 import iamutkarshtiwari.github.io.ananas.editimage.EditImageActivity;
 
 public abstract class BaseEditFragment extends Fragment {
+
     protected EditImageActivity activity;
 
-    protected EditImageActivity ensureEditActivity(){
-        if(activity == null){
+    private void ensureEditActivity() {
```
Comment (pal-tgiannakopoulos) [inline ananas/src/main/java/iamutkarshtiwari/github/io/ananas/editimage/fragment/BaseEditFragment.java]: make this protected and get rid of the field maybe?
it's causing a leak

File: ananas/src/main/java/iamutkarshtiwari/github/io/ananas/editimage/fragment/RotateFragment.java [hunk 6]
```diff
@@ -49,27 +53,24 @@ public void onCreate(Bundle savedInstanceState) {
...
-        loadingDialog = BaseActivity.getLoadingDialog(getActivity(), R.string.iamutkarshtiwari_github_io_ananas_loading,
-                false);
-        return mainView;
+        return inflater.inflate(R.layout.fragment_edit_image_rotate, null);
     }
 
     @Override
-    public void onActivityCreated(Bundle savedInstanceState) {
-        super.onActivityCreated(savedInstanceState);
+    public void onViewCreated(@NonNull @NotNull View view, @Nullable @org.jetbrains.annotations.Nullable Bundle savedInstanceState) {
```
Comment (pal-tgiannakopoulos) [inline ananas/src/main/java/iamutkarshtiwari/github/io/ananas/editimage/fragment/RotateFragment.java]: double imports with jetbrains nullable and non-not-null

File: ananas/src/main/java/iamutkarshtiwari/github/io/ananas/editimage/fragment/StickerFragment.java [hunk 7]
```diff
@@ -59,66 +57,63 @@ public void onCreate(Bundle savedInstanceState) {
...
-        return mainView;
+        return inflater.inflate(R.layout.fragment_edit_image_sticker_type, null);
     }
 
     @Override
-    public void onActivityCreated(Bundle savedInstanceState) {
-        super.onActivityCreated(savedInstanceState);
-        this.stickerView = activity.stickerView;
-        flipper = mainView.findViewById(R.id.flipper);
+    public void onViewCreated(@NonNull @NotNull View view, @Nullable @org.jetbrains.annotations.Nullable Bundle savedInstanceState) {
```
Comment (pal-tgiannakopoulos) [inline ananas/src/main/java/iamutkarshtiwari/github/io/ananas/editimage/fragment/StickerFragment.java]: same problem with imports

File: ananas/src/main/java/iamutkarshtiwari/github/io/ananas/editimage/fragment/StickerFragment.java [hunk 8]
```diff
@@ -59,66 +57,63 @@ public void onCreate(Bundle savedInstanceState) {
...
+        View backToMenu = view.findViewById(R.id.back_to_main);
         backToMenu.setOnClickListener(new BackToMenuClick());
 
-        View backToType = mainView.findViewById(R.id.back_to_type);
+        View backToType = view.findViewById(R.id.back_to_type);
         backToType.setOnClickListener(v -> flipper.showPrevious());
     }
 
+    @Override
+    public void onActivityCreated(Bundle savedInstanceState) {
```
Comment (pal-tgiannakopoulos) [inline ananas/src/main/java/iamutkarshtiwari/github/io/ananas/editimage/fragment/StickerFragment.java]: empty implementation

File: ananas/src/main/java/iamutkarshtiwari/github/io/ananas/editimage/fragment/crop/CropFragment.java [hunk 9]
```diff
@@ -128,16 +126,17 @@ private int getColorFromRes(@ColorRes int resId) {
     }
 
     @Override
-    public void onActivityCreated(Bundle savedInstanceState) {
-        super.onActivityCreated(savedInstanceState);
+    public void onViewCreated(@NonNull @NotNull View view, @Nullable @org.jetbrains.annotations.Nullable Bundle savedInstanceState) {
```
Comment (pal-tgiannakopoulos) [inline ananas/src/main/java/iamutkarshtiwari/github/io/ananas/editimage/fragment/crop/CropFragment.java]: same problem with imports

File: ananas/src/main/java/iamutkarshtiwari/github/io/ananas/editimage/fragment/paint/PaintFragment.kt [hunk 10]
```diff
@@ -0,0 +1,187 @@
...
+
+    fun savePaintImage() {
+        compositeDisposable.clear()
+        val applyPaintDisposable = applyPaint(activity.mainBit)
+                .flatMap { bitmap: Bitmap? ->
+                    if (bitmap == null) {
+                        return@flatMap Single.error<Bitmap>(Throwable("Error occurred while applying paint"))
+                    } else {
+                        return@flatMap Single.just(bitmap)
+                    }
```
Comment (pal-tgiannakopoulos) [inline ananas/src/main/java/iamutkarshtiwari/github/io/ananas/editimage/fragment/paint/PaintFragment.kt:129]: this seems to terminate the flatmap early in every case

File: ananas/src/main/java/iamutkarshtiwari/github/io/ananas/editimage/ImageEditorIntentBuilder.kt (hunk 1 above)
Comment (kavinprabhuc) [reply ananas/src/main/java/iamutkarshtiwari/github/io/ananas/editimage/ImageEditorIntentBuilder.kt]: If force portrait is not set the orientation will be locked for editor activity based on what orientation the device was. If user triggers image editor while the device is in landscape, then. image editor will be in landscape too.

File: ananas/src/main/java/iamutkarshtiwari/github/io/ananas/editimage/ImageEditorIntentBuilder.kt (hunk 2 above)
Comment (kavinprabhuc) [reply ananas/src/main/java/iamutkarshtiwari/github/io/ananas/editimage/ImageEditorIntentBuilder.kt]: Same as mentioned above

File: ananas/src/main/java/iamutkarshtiwari/github/io/ananas/editimage/ImageEditorIntentBuilder.kt (hunk 3 above)
Comment (kavinprabhuc) [reply ananas/src/main/java/iamutkarshtiwari/github/io/ananas/editimage/ImageEditorIntentBuilder.kt]: Done

File: ananas/src/main/java/iamutkarshtiwari/github/io/ananas/editimage/adapter/StickerAdapter.java (hunk 4 above)
Comment (kavinprabhuc) [reply ananas/src/main/java/iamutkarshtiwari/github/io/ananas/editimage/adapter/StickerAdapter.java]: Done

File: ananas/src/main/java/iamutkarshtiwari/github/io/ananas/editimage/fragment/BaseEditFragment.java (hunk 5 above)
Comment (kavinprabhuc) [reply ananas/src/main/java/iamutkarshtiwari/github/io/ananas/editimage/fragment/BaseEditFragment.java]: Done. Also updated all the affected places with not null checks

File: ananas/src/main/java/iamutkarshtiwari/github/io/ananas/editimage/fragment/RotateFragment.java (hunk 6 above)
Comment (kavinprabhuc) [reply ananas/src/main/java/iamutkarshtiwari/github/io/ananas/editimage/fragment/RotateFragment.java]: Fixed

File: ananas/src/main/java/iamutkarshtiwari/github/io/ananas/editimage/fragment/StickerFragment.java (hunk 7 above)
Comment (kavinprabhuc) [reply ananas/src/main/java/iamutkarshtiwari/github/io/ananas/editimage/fragment/StickerFragment.java]: Fixed

File: ananas/src/main/java/iamutkarshtiwari/github/io/ananas/editimage/fragment/StickerFragment.java (hunk 8 above)
Comment (kavinprabhuc) [reply ananas/src/main/java/iamutkarshtiwari/github/io/ananas/editimage/fragment/StickerFragment.java]: Removed

File: ananas/src/main/java/iamutkarshtiwari/github/io/ananas/editimage/fragment/crop/CropFragment.java (hunk 9 above)
Comment (kavinprabhuc) [reply ananas/src/main/java/iamutkarshtiwari/github/io/ananas/editimage/fragment/crop/CropFragment.java]: Fixed

File: ananas/src/main/java/iamutkarshtiwari/github/io/ananas/editimage/fragment/paint/PaintFragment.kt (hunk 10 above)
Comment (kavinprabhuc) [reply ananas/src/main/java/iamutkarshtiwari/github/io/ananas/editimage/fragment/paint/PaintFragment.kt:129]: This seems to be working fine. We had to return the bitmap or the error

File: ananas/src/main/java/iamutkarshtiwari/github/io/ananas/editimage/EditImageActivity.java [hunk 11]
```diff
@@ -450,30 +452,7 @@ protected void onDestroy() {
     protected void setLockScreenOrientation(boolean lock) {
         if (Build.VERSION.SDK_INT >= 18) {
             setRequestedOrientation(lock ? ActivityInfo.SCREEN_ORIENTATION_LOCKED : ActivityInfo.SCREEN_ORIENTATION_FULL_SENSOR);
-            return;
         }
-
-        if (lock) {
```
Comment (pal-tgiannakopoulos) [inline ananas/src/main/java/iamutkarshtiwari/github/io/ananas/editimage/EditImageActivity.java:456]: ok this will not create a problem for us since we will never run this on lower than 18 but I think that it is wrong to completely remove these.
You could change the pattern with an early return though and put an else on the if in line 451 but please keep this code for lower platforms as it will makes sense if we want to do a PR to contribute back to the open-source project

File: ananas/src/main/java/iamutkarshtiwari/github/io/ananas/editimage/ImageEditorIntentBuilder.kt (hunk 1 above)
Comment (pal-tgiannakopoulos) [reply ananas/src/main/java/iamutkarshtiwari/github/io/ananas/editimage/ImageEditorIntentBuilder.kt]: ok, in this case then let's keep the forcePortrait(true) as you initially did

File: ananas/src/main/java/iamutkarshtiwari/github/io/ananas/editimage/fragment/BaseEditFragment.java (hunk 5 above)
Comment (pal-tgiannakopoulos) [reply ananas/src/main/java/iamutkarshtiwari/github/io/ananas/editimage/fragment/BaseEditFragment.java]: As talked in chat, we should hold a local variable and not calling the new method all the time

File: ananas/src/main/java/iamutkarshtiwari/github/io/ananas/editimage/fragment/RotateFragment.java [hunk 12]
```diff
@@ -119,79 +114,100 @@ public void onClick(View v) {
...
+                        .observeOn(AndroidSchedulers.mainThread())
+                        .doOnSubscribe(subscriber -> {
+                            EditImageActivity activityInstance;
+                            if ((activityInstance = getActivityInstance()) != null)
+                                activityInstance.showLoadingDialog();
+                        })
+                        .doFinally(() -> {
+                            EditImageActivity activityInstance;
+                            if ((activityInstance = getActivityInstance()) != null)
+                                activityInstance.showLoadingDialog();
```
Comment (pal-tgiannakopoulos) [inline ananas/src/main/java/iamutkarshtiwari/github/io/ananas/editimage/fragment/RotateFragment.java]: is this change deliberate?
before this change the "doFinally" was running dismiss on the dialog, not show

Comment (kavinprabhuc) [reply ananas/src/main/java/iamutkarshtiwari/github/io/ananas/editimage/fragment/RotateFragment.java]: It should be dismiss, I'll fix and update the same

**Review by pal-tgiannakopoulos — APPROVED** [review]
